tq/ssh.go

```diff
@@ -0,0 +1,405 @@
+	ctx      lfshttp.Context
+	transfer *ssh.SSHTransfer
+}
+
+// WorkerStarting is called when a worker goroutine starts to process jobs
+// Implementations can run some startup logic here & return some context if needed
+func (a *SSHAdapter) WorkerStarting(workerNum int) (interface{}, error) {
+	a.transfer.SetConnectionCountAtLeast(workerNum + 1)
+	return a.transfer.Connection(workerNum), nil
+}
+
+// WorkerEnding is called when a worker goroutine is shutting down
+// Implementations can clean up per-worker resources here, context is as returned from WorkerStarting
+func (a *SSHAdapter) WorkerEnding(workerNum int, ctx interface{}) {
+}
+
+func (a *SSHAdapter) tempDir() string {
+	// Shared with the basic download adapter.
+	d := filepath.Join(a.fs.LFSStorageDir, "incomplete")
+	if err := tools.MkdirAll(d, a.fs); err != nil {
+		return os.TempDir()
+	}
+	return d
+}
+
+// DoTransfer performs a single transfer within a worker. ctx is any context returned from WorkerStarting
+func (a *SSHAdapter) DoTransfer(ctx interface{}, t *Transfer, cb ProgressCallback, authOkFunc func()) error {
+	if authOkFunc != nil {
+		authOkFunc()
+	}
+	conn := ctx.(*ssh.PktlineConnection)
+	if a.adapterBase.direction == Upload {
+		return a.upload(t, conn, cb)
+	} else {
+		return a.download(t, conn, cb)
+	}
+}
+
+func (a *SSHAdapter) download(t *Transfer, conn *ssh.PktlineConnection, cb ProgressCallback) error {
+	rel, err := t.Rel("download")
+	if err != nil {
+		return err
+	}
+	if rel == nil {
+		return errors.Errorf("No download action for object: %s", t.Oid)
+	}
+	// Reserve a temporary filename. We need to make sure nobody operates on the file simultaneously with us.
+	f, err := tools.TempFile(a.tempDir(), t.Oid, a.fs)
+	if err != nil {
+		return err
+	}
+	tmpName := f.Name()
+	defer func() {
+		if f != nil {
+			f.Close()
+		}
+		os.Remove(tmpName)
+	}()
+
+	return a.doDownload(t, conn, f, cb)
+}
+
+// doDownload starts a download. f is expected to be an existing file open in RW mode
+func (a *SSHAdapter) doDownload(t *Transfer, conn *ssh.PktlineConnection, f *os.File, cb ProgressCallback) error {
+	args := a.argumentsForTransfer(t, "download")
+	conn.Lock()
+	defer conn.Unlock()
+	err := conn.SendMessage(fmt.Sprintf("get-object %s", t.Oid), args)
+	if err != nil {
+		return err
+	}
+	status, args, data, err := conn.ReadStatusWithData()
+	if err != nil {
+		return err
+	}
+	if status < 200 || status > 299 {
+		buffer := &bytes.Buffer{}
+		if data != nil {
+			io.Copy(buffer, data)
+		}
+		if buffer.Len() > 1024 {
+			buffer.Truncate(1024)
+		}
+		return errors.NewRetriableError(fmt.Errorf("got status %d when fetching OID %s: %s", status, t.Oid, buffer.String()))
+	}
+
+	actualSize := int64(-1)
+	for _, arg := range args {
+		entries := strings.SplitN(arg, "=", 2)
+		if entries[0] == "size" {
+			val, err := strconv.ParseInt(entries[1], 10, 64)
+			if err != nil {
+				return fmt.Errorf("expected valid size, got %q", entries[1])
+			}
+			actualSize = val
+		}
+	}
```

Comment:
With my "bad server" thinking cap on again, should we guard against multiple or invalid (e.g., negative) `size` argument packet lines here?

The proposed APi spec does [say](https://github.com/git-lfs/git-lfs/blob/05f7e0c8b8ee7378ebbb87b8fb5d655a871cf2b3/docs/proposals/ssh_adapter.md#requests-to-transfer-objects) that "unknown arguments should be ignored".  But since the `"size"` argument is known and expected, maybe something like the following would cover cases of extra, missing, and invalid ones?  (I'm unsure if a zero size is acceptable; I allowed it here.)

```suggestion
	var actualSize int64
	seenSize := false
	for _, arg := range args {
		if strings.HasPrefix(arg, "size=") {
			if seenSize {
				return errors.NewProtocolError("unexpected size argument", nil)
			}
			actualSize, err = strconv.ParseInt(arg[5:], 10, 64)
			if err != nil || actualSize < 0 {
				return errors.NewProtocolError(fmt.Sprintf("expected valid size, got %q", arg[5:]), err)
			}
			seenSize = true
		}
	}
	if !seenSize {
		return errors.NewProtocolError("no size argument seen", nil)
	}
```

Reply:
Sure.  I've squashed in your changes with a co-author credit.